### PR TITLE
Handbook: move tutorials and reference docs to the top-level

### LINKED ISSUE #13962: Handbook: Improve navigation on the sidebar
For me as a developer it's currently quite difficult to use the handbook/documentation. The sidebar is way too long and it's pretty hard to find what I need.

I think that only the direct child items should be visible by default and when I click on one of them, it will be expanded. Just like it already behaves with "Designer & Developer Handbook" and "Contributors Guide".

Link to the documentation: [https://wordpress.org/gutenberg/handbook/designers-developers/](https://wordpress.org/gutenberg/handbook/designers-developers/)

Developer Documentation Screenshot

### PR DESCRIPTION
**Pull request by oandregal**
Fixes https://github.com/WordPress/gutenberg/issues/13962

After multiple iterations by many people, the Gutenberg handbook has a lot of high-quality content :sparkling_heart: :scroll: Its biggest issue right now is discoverability. This proposal seeks to improve that while adhering to the current constraints (sidebar only collapsing for the top-level sections, etc.).

In essence, it moves up the tutorials and the reference documentation to the top-level. This would be the TOC and the sidebar:

- [Gutenberg Handbook](https://wordpress.org/gutenberg/handbook/)
- [Designer & Developer Handbook](https://wordpress.org/gutenberg/handbook/designers-developers/)
- [Contributors Guide](https://wordpress.org/gutenberg/handbook/contributors/)
- [Tutorials](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/)
- [Block API Reference](https://wordpress.org/gutenberg/handbook/designers-developers/developers/block-api/)
- [Component Reference](https://wordpress.org/gutenberg/handbook/designers-developers/developers/components/)
- [Data Module Reference](https://wordpress.org/gutenberg/handbook/designers-developers/developers/data/)
- [Filter Reference](https://wordpress.org/gutenberg/handbook/designers-developers/developers/filters/)
- [Package Reference](https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/)

It could be argued that reference docs could be grouped within a "Reference" section. I could agree with that sentiment. However, that approach would suffer from the same problems this PR is trying to fix (long sidebar, lack of discoverability), so I'd rather have them at the top-level.

### CODE REVIEW

File: docs/data/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Data Module Reference
```
Comment (nosolosw) [inline docs/data/README.md]: This file was previously: `docs/designers-developers/developers/data/README.md`. It looks like GitHub didn't pick up this rename.

File: docs/designers-developers/developers/data/README.md [hunk 2]
```diff
@@ -1,11 +0,0 @@
-# Data Module Reference
```
Comment (nosolosw) [inline docs/designers-developers/developers/data/README.md]: Moved to `docs/data/README.md`. It looks like GitHub didn't pick up this rename.

File: docs/designers-developers/developers/tutorials/readme.md [hunk 3]
```diff
@@ -1,13 +0,0 @@
-# Tutorials
```
Comment (nosolosw) [inline docs/designers-developers/developers/tutorials/readme.md]: Moved to `docs/tutorials/README.md`. It looks like GitHub didn't pick up this rename.

File: docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-0.md [hunk 4]
```diff
@@ -1,12 +0,0 @@
-# Creating a Sidebar for Your Plugin
```
Comment (nosolosw) [inline docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-0.md]: Moved to `docs/tutorials/sidebar-tutorial/plugin-sidebar-0.md`. It looks like GitHub didn't pick up this rename.

File: docs/manifest.json [hunk 5]
```diff
@@ -161,353 +71,443 @@
...
+		"slug": "block-annotations",
+		"markdown_source": "https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/block-api/block-annotations.md",
+		"parent": "block-api"
 	},
 	{
-		"title": "Gutenberg Release Process",
-		"slug": "release",
-		"markdown_source": "https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/contributors/release.md",
-		"parent": "develop"
+		"title": "Components",
```
Comment (nosolosw) [inline docs/manifest.json]: I see that components subsections are added at the end, not after this (same for Data module subsections). Is that relevant? My understanding is that it isn't. cc @pento @youknowriad @chrisvanpatten @mkaz

File: docs/tutorials/readme.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# Tutorials
```
Comment (nosolosw) [inline docs/tutorials/readme.md]: This file was previously: `docs/designers-developers/developers/tutorials/README.md`. It looks like GitHub didn't pick up this rename.

File: docs/tutorials/sidebar-tutorial/plugin-sidebar-0.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# Creating a Sidebar for Your Plugin
```
Comment (nosolosw) [inline docs/tutorials/sidebar-tutorial/plugin-sidebar-0.md]: This file was previously: `docs/designers-developers/developers/tutorials/sidebar-tutorial/plugin-sidebar-0.md`. It looks like GitHub didn't pick up this rename.

**Comment by gziolo** [comment]
What about all the existing url references from blog posts, external projects, search engines, etc.?

**Comment by nosolosw** [comment]
> What about all the existing url references from blog posts, external projects, search engines, etc.?

That's a good question. How do we proceed in cases like this? Is it possible to set up redirects?

**Comment by gziolo** [comment]
@chrisvanpatten should know better as he led the previous handbook structure redesign. 

What if we move all auto-generated API docs to their own subfolder instead? I think this would make the menu easier to scan. In particular, I think about:

- https://wordpress.org/gutenberg/handbook/designers-developers/developers/data/
- https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/
- https://wordpress.org/gutenberg/handbook/designers-developers/developers/components/

I also think that tutorials should be one level up:
- https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/ -> https://wordpress.org/gutenberg/handbook/designers-developers/tutorials/

In addition, if we make all sections in the menu collapsed by default, it should make everything way much easier to discover.

**Comment by nosolosw** [comment]
@gziolo My understanding on how the handbook works is that the sidebar only collapses for top-level sections. If that assumption holds true, the only approach I see to fix the sidebar length is to move the longer sections to the top-level (what this PR proposes). Any other approach will suffer from the same lack of discoverability.

Note that I'm not interested in re-architecting the structure of the handbook! I see this proposal as a workaround (that makes sense) to fix the bigger problem we have (improve discoverability).

**Comment by gziolo** [comment]
> My understanding on how the handbook works is that the sidebar only collapses for top-level sections. If that assumption holds true, the only approach I see to fix the sidebar length is to move the longer sections to the top-level (what this PR proposes).

Yes, the sidebar is broken by design :) We should make it possible to collapse subitems for inactive branches rather than play with URL or folder structure. it works perfectly fine for top-level items, so I can't think of any reason why it would be not possible for nested sections.

**Comment by mkaz** [comment]
Redirects are possible, but I'm not sure how. I see the request occasional made in #meta 

This PR overall highlights the biggest issue we have with the documentation, we are publishing into a system we have limited control over, from server config to design. 

An even more radical PR approach would be setting up a separate server to publish into, something like Github Pages might be better for documentation. I'm not sure if the goal of getting all documentation into a WordPress instance is all that worthy. This makes it even harder to contribute since users need to be allowed access.

Plus this creates a disconnect, the theme & design are disconnected and not updatable with the documentation. So a contributor who sees an issue with something design, or layout related can't contribute for example the issue we have with the sidebar.

**Comment by chrisvanpatten** [comment]
Hi all! 👋  I have many things to add!

> @gziolo:
> We should make it possible to collapse subitems for inactive branches rather than play with URL or folder structure. it works perfectly fine for top-level items, so I can't think of any reason why it would be not possible for nested sections.

I agree completely with this, however with the caveat that many folks have tried desperately to find a way to fix the handbook theme, and all have failed.

It would be really helpful of someone with more history/at a higher level in the project (@mtias? @karmatosed?) _who knows who originally set up the handbook_ could point that person to this ticket, so we can talk to them about getting the handbook environment fixed. There are a lot of other UX/UI issues, not just the menu, that could be resolved in the same go.

At this point it's not at all clear who set this up, which Trac repository holds the handbook code (if any?), what the contribution process is, etc.

> @mkaz 
> This PR overall highlights the biggest issue we have with the documentation, we are publishing into a system we have limited control over, from server config to design.

Yes, very much this :)

> @nosolosw 
> How do we proceed in cases like this? Is it possible to set up redirects?

This was a big problem when we did the first reorganization. It also causes problems with moving around the files because the markdown sync plugin isn't always smart about moving pages, and sometimes you can end up with duplicate pages.

> @mkaz 
> An even more radical PR approach would be setting up a separate server to publish into, something like Github Pages might be better for documentation. I'm not sure if the goal of getting all documentation into a WordPress instance is all that worthy. This makes it even harder to contribute since users need to be allowed access.

Ultimately I think this is a really important point, but I want to also toss in the wrench that the awesome WordPress docs team, lead by @Kenshino, has done a TON of work setting up DevHub and HelpHub.

I would actually argue an even better solution would be partnering up with @Kenshino, and anyone with available time on the Docs team, to build a _totally new importer_ that can push this content directly into DevHub. Then the Gutenberg handbook site could be completely deprecated and redirected to DevHub.

@danielbachhuber Might also have some context on how WP-CLI's documentation process works (and who helped build it), which I think is the closest analogue to our ideal flow with the Gutenberg documentation.

**Comment by nosolosw** [comment]
@gziolo @mkaz @chrisvanpatten I understand those frustrations and I think those ideas make sense. However, they are big undertakings that won't be built overnight. In the meantime, I'd like the WordPress 5.2 release to highlight all the amount of work that went into docs during the past months.

I'll look into what we can do with the redirects and report back.

Provided we can solve that, I truly believe these changes make sense: they make more visible the things people usually look for (tutorials and reference/API docs).

**Comment by mkaz** [comment]
@nosolosw  agreed, my idea was more future thinking than related to this PR. 

If we can get redirects in place, I'm :+1: on this PR. My goal was to bring up some ideas worth noodling about how to solve in the future.

**Comment by nosolosw** [comment]
@mkaz Agreed, and I think it is valuable to bounce those ideas. One thing I'm very excited about and look forward to is the Playground Riad has bootstrapped recently. It could be all the things you mentioned. In the meantime, though, we need to fix with what have :shrug:

**Comment by chrisvanpatten** [comment]
@nosolosw I would check in with @dd32; he helped a bunch the first time we reorganized the docs.

**Comment by Kenshino** [comment]
Hey all - Jon from the Documentation Team here!

I've discussed with @chrisvanpatten and @0aveRyan a few times about how we should be structuring documentation 
I'm not digging in deep about table of contents but really where all documentation (including the block editor) should reside.

The general mandate of us making and maintaining documentation at WordPress.org is to ensure that there is only one specific place for developer documentation - which is DevHub (developer.wordpress.org)

So it's always been assumed and discussed that it would be in e.g. https://developer.wordpress.org/block-editor

Screenshot 2019-03-29 00 19 06

We've trained WordPress developers for years to go into DevHub (Codex is actively being deprecated). So we should really capitalise on this trained behaviour.

Beyond where it should reside - I know we're also discussing the technology syndicating it and how it would be maintained.

I'd really like to encourage the Gutenberg team to consider what the WP CLI team did

- Work within the typical handbook CPT
- Copy paste content over 
- Work out the technology to syndicate after (they used a modified version of RESTPLAIN to syndicate from GitHub to DevHub)

What should go into DevHub's Handbook are the "How do you develop with Gutenberg" docs. i.e. All the block tutorials you have such as 'Writing your first block type'

What should go into the Code Reference are all the actual source code docs which I believe is going to be resolved with the JS code parser.

You already have a wealth of information here (it's crazy good) - and the key focus here I'd like everyone to look at is

1. Some documentation is better than no documentation
2. Documentation in the right place is more important than the technology that powers it. Discoverability only makes sense if people know where to go to :)

**Comment by danielbachhuber** [comment]
> @danielbachhuber Might also have some context on how WP-CLI's documentation process works (and who helped build it), which I think is the closest analogue to our ideal flow with the Gutenberg documentation.

Are there some specific questions I can help answer?

WP-CLI has a dedicated handbook repository: https://github.com/wp-cli/handbook The markdown files are generated with a series of scripts https://github.com/wp-cli/handbook/tree/master/bin and then imported into make.wordpress.org/cli/handbook.

We avoid the codebase documentation discrepancy issue by only regenerating docs at the time of release. Tutorials, etc. can be updated whenever.

**Comment by chrisvanpatten** [comment]
@danielbachhuber It's mainly this bit that I'm curious about:

> then imported into make.wordpress.org/cli/handbook

A few questions:

+ How does that import happen?
+ Is the code that runs the import open source somewhere?
+ Does someone have to trigger the import, or is it scheduled in some way?
+ Who wrote the code that runs the import, and who currently "owns" it?
+ Are there any issues with the workflow that you've encountered?

Thanks 🙌

**Comment by danielbachhuber** [comment]
> * How does that import happen?
> * Is the code that runs the import open source somewhere?

Using some code here: https://meta.trac.wordpress.org/browser/sites/trunk/wordpress.org/public_html/wp-content/plugins/wporg-cli/inc/class-markdown-import.php

In fact, the existing Gutenberg handbook uses a derivative of this!

> * Does someone have to trigger the import, or is it scheduled in some way?

WP Cron for lyfe!

> * Who wrote the code that runs the import, and who currently "owns" it?

I wrote it. Everyone owns the copyright!

> * Are there any issues with the workflow that you've encountered?

Seems to work reasonably well, although @schlessera might have opinions about it.

Personally, I much prefer to have the codebase documentation updated at release time, over being continuously updated. I think it's a reasonable expectation the published documentation correlates with the latest stable release.

File: docs/manifest.json [hunk 8]
```diff
@@ -161,353 +71,443 @@
...
+		"parent": "contributors"
+	},
+	{
+		"title": "Outreach",
+		"slug": "outreach",
+		"markdown_source": "https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/contributors/outreach.md",
+		"parent": "contributors"
+	},
 	{
 		"title": "Tutorials",
```
Comment (nosolosw) [inline docs/manifest.json]: This is what @dd32 taught me about redirects:

- The code for the redirects lives [here](https://meta.trac.wordpress.org/browser/sites/trunk/wordpress.org/public_html/wp-content/themes/pub/gutenberg/functions.php?marks=856#L850).
- It is possible to set up manual redirects (see code above).
- If the slug of the page (last bit in `/this/is/the/SLUG/`) remains the same it'll be redirected automatically.

The changes introduced in this PR shorten the page URL removing the `designers-developers/developers` part, so my understanding is that they'll be redirected automatically:

Section | Before | After
--- | --- | ---
Tutorials (+33 subsections) | https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/ | https://wordpress.org/gutenberg/handbook/tutorials/
Block API (+6 subsections) | https://wordpress.org/gutenberg/handbook/designers-developers/developers/block-api/ | https://wordpress.org/gutenberg/handbook/block-api/
Components (+65 subsections) | https://wordpress.org/gutenberg/handbook/designers-developers/developers/components/ | https://wordpress.org/gutenberg/handbook/components/
Data (+9 subsections) | https://wordpress.org/gutenberg/handbook/designers-developers/developers/data/ | https://wordpress.org/gutenberg/handbook/data/
Filters (+4 subsections) | https://wordpress.org/gutenberg/handbook/designers-developers/developers/filters/ | https://wordpress.org/gutenberg/handbook/filters/
Packages (+19 subsections) | https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/ | https://wordpress.org/gutenberg/handbook/packages/

File: docs/manifest.json (hunk 5 above)
Comment (nosolosw) [reply docs/manifest.json]: The order of sections in the handbook determines the prev/next links. See, for example the nex/prev links of Outreach (the last section of the hanbook as per the sidebar):

outreach-prev-next-links

Note how the next link points to the first package.

Created an issue for this https://github.com/WordPress/gutenberg/issues/14794

**Comment by nosolosw** [comment]
This is where things stand for this PR:

- The changes here are ready. There is wide agreement that this fixes our current handbook for users, which is a frustrating experience as it stands. There is also interest in bigger changes that I'd like to tackle independently so we're not blocked to improve what we currently have.

- The old URLs will be automatically redirected to the new ones. If a particular section fails to do this, we can add the manual redirect after. I'll follow up on this.

All things considered, I'm prepared to merge this in as soon as someone gives me a :+1: Would like to time the merge to make sure someone from meta is around.

File: docs/manifest.json (hunk 8 above)
Comment (nosolosw) [reply docs/manifest.json]: Got confirmation from Dion that this holds true. If any particular section fails to redirect we can fix it in a case by case basis. We don't expect any to fail, though, and I've tested that the proposed changes are already redirected.

**Comment by nosolosw** [comment]
Relevant discussion in [this Slack thread](https://wordpress.slack.com/archives/C02RP4WU5/p1554732395115000).

**Comment by nosolosw** [comment]
Closing this in favor of https://meta.trac.wordpress.org/ticket/4388

**Pull request closed without merging**